# Assignment - 7

# Salary Discrimination

The file `salary.txt` contains data for 52 tenure-track professors at a small Midwestern college. This data was used in legal proceedings in the 1980s about discrimination against women in salary.

The data in the file, by column:

1. Sex. 1 for female, 0 for male.
2. Rank. 1 for assistant professor, 2 for associate professor, 3 for full professor.
3. Year. Number of years in current rank.
4. Degree. Highest degree. 1 for doctorate, 0 for master's.
5. YSdeg. Years since highest degree was earned.
6. Salary. Salary/year in dollars.

## Import Libraries

In [55]:
import numpy as np
import pandas as pd
from sklearn.linear_model import LinearRegression
import matplotlib.pyplot as plt
import itertools as it
import seaborn as sns

## Tasks

1. Find the linear regression equation for this data using columns 1-5 to column 6.

In [3]:
df = pd.read_fwf("salary.txt", header=None, 
                 names=["Sex", "Rank", "Year", "Degree", "YSdeg", "Salary"])
df.head()

Sex  Rank  Year  Degree  YSdeg  Salary
0    0     3    25       1     35   36350
1    0     3    13       1     22   35350
2    0     3    10       1     23   28200
3    1     3     7       1     27   26775
4    0     3    19       0     30   33696

In [6]:
X_train = df.drop('Salary', axis=1)
Y_train = df['Salary'].values.reshape(-1,1)

# Linear Equation:
reg_LR = LinearRegression()
reg_LR.fit(X_train, Y_train)
Y_predict = reg_LR.predict(X_train)

2. Find the selection of columns with the best $R^2$ score.

In [51]:
col_combos = []

for i in range(len(df.columns)):
    col_combos += list(it.combinations(X_train.columns, i+1))

col_combos

[('Sex',),
 ('Rank',),
 ('Year',),
 ('Degree',),
 ('YSdeg',),
 ('Sex', 'Rank'),
 ('Sex', 'Year'),
 ('Sex', 'Degree'),
 ('Sex', 'YSdeg'),
 ('Rank', 'Year'),
 ('Rank', 'Degree'),
 ('Rank', 'YSdeg'),
 ('Year', 'Degree'),
 ('Year', 'YSdeg'),
 ('Degree', 'YSdeg'),
 ('Sex', 'Rank', 'Year'),
 ('Sex', 'Rank', 'Degree'),
 ('Sex', 'Rank', 'YSdeg'),
 ('Sex', 'Year', 'Degree'),
 ('Sex', 'Year', 'YSdeg'),
 ('Sex', 'Degree', 'YSdeg'),
 ('Rank', 'Year', 'Degree'),
 ('Rank', 'Year', 'YSdeg'),
 ('Rank', 'Degree', 'YSdeg'),
 ('Year', 'Degree', 'YSdeg'),
 ('Sex', 'Rank', 'Year', 'Degree'),
 ('Sex', 'Rank', 'Year', 'YSdeg'),
 ('Sex', 'Rank', 'Degree', 'YSdeg'),
 ('Sex', 'Year', 'Degree', 'YSdeg'),
 ('Rank', 'Year', 'Degree', 'YSdeg'),
 ('Sex', 'Rank', 'Year', 'Degree', 'YSdeg')]

In [52]:
def reg_score(input_cols):
    X_train = df[list(input_cols)]
    Y_train = df['Salary'].values.reshape(-1,1)

    # Linear Equation:
    reg_LR = LinearRegression()
    reg_LR.fit(X_train, Y_train)
    return reg_LR.score(X_train, Y_train)

In [53]:
max_score = 0
trains = []
for i in col_combos:
    score = reg_score(i)
    if max_score < score:
        trains = []
        max_score = score
        trains += list(i)
    print(i, reg_score(i))

('Sex',) 0.06389892583288015
('Rank',) 0.7525360539274633
('Year',) 0.49093702676921624
('Degree',) 0.004861680984748373
('YSdeg',) 0.45542813458360276
('Sex', 'Rank') 0.7554917749443394
('Sex', 'Year') 0.4911360941886387
('Sex', 'Degree') 0.06643739842018881
('Sex', 'YSdeg') 0.4930479271966782
('Rank', 'Year') 0.8436480933264598
('Rank', 'Degree') 0.7557961688386643
('Rank', 'YSdeg') 0.7623948189698249
('Year', 'Degree') 0.4919014149016442
('Year', 'YSdeg') 0.5782019226431887
('Degree', 'YSdeg') 0.5409662288237875
('Sex', 'Rank', 'Year') 0.8454319861987166
('Sex', 'Rank', 'Degree') 0.7583144768153612
('Sex', 'Rank', 'YSdeg') 0.7666122610227071
('Sex', 'Year', 'Degree') 0.49207944757472966
('Sex', 'Year', 'YSdeg') 0.5808063918180405
('Sex', 'Degree', 'YSdeg') 0.5832708257660315
('Rank', 'Year', 'Degree') 0.8437395200874137
('Rank', 'Year', 'YSdeg') 0.8456725764485598
('Rank', 'Degree', 'YSdeg') 0.7625149185326454
('Year', 'Degree', 'YSdeg') 0.6236261611533505
('Sex', 'Rank', 'Year', 'D

In [54]:
print(trains, max_score) # Best R2 Score

['Sex', 'Rank', 'Year', 'Degree', 'YSdeg'] 0.8547180674410969


3. Report whether sex is a factor in salary.

R2 Score for ('Sex', 'Salary') is 0.06389892583288015 which indicates no direct influence in salary.